File: design-patterns/go/iterator/pagination/iterator.go
// iterator.go

package main

type Iterator struct {
	currentPosition int
	pages           []Page
}

func NewIterator(pages []Page) (iterator *Iterator) {
	iterator = &Iterator{}
	iterator.currentPosition = 0
	iterator.pages = pages
	return
}

func (iterator *Iterator) HasNext() bool {
	return iterator.currentPosition < len(iterator.pages)
}

func (iterator *Iterator) Next() Page {
	page := iterator.pages[iterator.currentPosition]
	iterator.currentPosition++
	return page
}

File: design-patterns/ts/iterator/pagination/iterator.ts
// iterator.ts

import AbstractIterator from "./abstract-iterator";
import Page from "./page";

class Iterator implements AbstractIterator {
    private currentPosition: number = 0;
    private readonly pages = new Map<number, Page>();

    constructor(pages: Map<number, Page>) {
        this.pages = pages;
    }

    hasNext(): boolean {
        return this.currentPosition < this.pages.size;
    }

    next(): Page | undefined {
        const page = this.pages.get(this.currentPosition);
        this.currentPosition++;

        return page;
    }
}

export default Iterator;